### PR TITLE
Upgrade etcd-rs to 0.3.0

### PR DESCRIPTION
**Pull request by mrwwheat7**

### CODE REVIEW

File: src/etcd_client.rs [hunk 1]
```diff
@@ -20,14 +19,9 @@ impl EtcdClient {
             etcd_rs::Client::connect(etcd_rs::ClientConfig {
                 endpoints: etcd_address_vec.clone(),
                 auth: None,
+                tls: None,
             })
             .await
```
Comment (pwang7) [inline src/etcd_client.rs:25]: add `.context(format!("failed to build etcd client to addresses={:?}", etcd_address_vec))`

File: src/etcd_client.rs [hunk 2]
```diff
@@ -53,21 +45,14 @@ impl EtcdClient {
...
         value: &T,
     ) -> anyhow::Result<Option<T>> {
-        let bin_value = bincode::serialize(value).map_err(|e| {
-            anyhow!(
-                "failed to encode {:?} to binary, the error is: {}",
-                value,
-                e,
-            )
-        })?;
+        let bin_value = bincode::serialize(value)?;
```
Comment (pwang7) [inline src/etcd_client.rs]: add `.context(format!("failed to encode {:?} to binary", value))?;`

File: src/etcd_client.rs [hunk 3]
```diff
@@ -36,9 +30,7 @@ impl EtcdClient {
     async fn get_list_async<T: DeserializeOwned>(&self, prefix: &str) -> anyhow::Result<Vec<T>> {
         let req = etcd_rs::RangeRequest::new(etcd_rs::KeyRange::prefix(prefix));
         let mut resp =
-            self.etcd_rs_client.kv().range(req).await.map_err(|e| {
-                anyhow!("failed to get RangeResponse from etcd, the error is: {}", e)
-            })?;
+            self.etcd_rs_client.kv().range(req).await?;
```
Comment (pwang7) [inline src/etcd_client.rs]: add `.context(format!("failed to get RangeResponse from etcd for RangeRequest={:?}", req))?;`

File: src/etcd_client.rs [hunk 4]
```diff
@@ -187,7 +167,6 @@ impl EtcdClient {
                 .kv()
                 .delete(req)
                 .await
-                .map_err(|e| anyhow!("failed to clear all data from etcd, the error is: {}", e,))
         })?;
```
Comment (pwang7) [inline src/etcd_client.rs]: add `.context("failed to clear all data from etcd")`

File: src/etcd_client.rs [hunk 5]
```diff
@@ -83,12 +68,7 @@ impl EtcdClient {
     ) -> anyhow::Result<Vec<T>> {
         let mut req = etcd_rs::DeleteRequest::new(etcd_rs::KeyRange::key(key));
         req.set_prev_kv(true);
-        let mut resp = self.etcd_rs_client.kv().delete(req).await.map_err(|e| {
-            anyhow!(
-                "failed to get DeleteResponse from etcd, the error is: {}",
-                e,
-            )
-        })?;
+        let mut resp = self.etcd_rs_client.kv().delete(req).await?;
```
Comment (pwang7) [inline src/etcd_client.rs]: add `.context(format!("failed to get DeleteResponse from etcd for DeleteRequest={:?}", req))?;`

File: src/etcd_client.rs [hunk 6]
```diff
@@ -53,21 +45,14 @@ impl EtcdClient {
...
+        let bin_value = bincode::serialize(value)?;
         let mut req = etcd_rs::PutRequest::new(key, bin_value);
         req.set_prev_kv(true); // Return previous value
         let mut resp = self
             .etcd_rs_client
             .kv()
             .put(req)
-            .await
-            .map_err(|e| anyhow!("failed to get PutResponse from etcd, the error is: {}", e))?;
+            .await?;
```
Comment (pwang7) [inline src/etcd_client.rs]: add `.context(format!("failed to get PutResponse from etcd for PutRequest={:?}", req))?;`

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: src/etcd_client.rs (hunk 3 above)
Comment (mrwwheat7) [reply src/etcd_client.rs]: The request doesn't implement `std::fmt::Debug`

File: src/etcd_client.rs [hunk 7]
```diff
@@ -20,25 +20,24 @@ impl EtcdClient {
...
             .await
-            .map_err(|e| {
-                anyhow!(format!(
-                    "failed to build etcd client to {:?}, the error is: {}",
-                    etcd_address_vec, e,
-                ))
-            })
+            .context(format!(
+                "failed to build etcd client to addresses={:?}",
+                etcd_address_vec
```
Comment (pwang7) [inline src/etcd_client.rs]: add trailing comma `etcd_address_vec,`

File: src/etcd_client.rs [hunk 8]
```diff
@@ -20,25 +20,24 @@ impl EtcdClient {
...
     /// Get key-value list from etcd
     async fn get_list_async<T: DeserializeOwned>(&self, prefix: &str) -> anyhow::Result<Vec<T>> {
         let req = etcd_rs::RangeRequest::new(etcd_rs::KeyRange::prefix(prefix));
-        let mut resp =
-            self.etcd_rs_client.kv().range(req).await.map_err(|e| {
-                anyhow!("failed to get RangeResponse from etcd, the error is: {}", e)
-            })?;
+        let mut resp = self.etcd_rs_client.kv().range(req).await.context(format!(
+            "failed to get RangeResponse from etcd for key={:?}",
+            prefix
```
Comment (pwang7) [inline src/etcd_client.rs]: add trailing comma

File: src/etcd_client.rs [hunk 9]
```diff
@@ -53,21 +52,14 @@ impl EtcdClient {
...
         req.set_prev_kv(true); // Return previous value
-        let mut resp = self
-            .etcd_rs_client
-            .kv()
-            .put(req)
-            .await
-            .map_err(|e| anyhow!("failed to get PutResponse from etcd, the error is: {}", e))?;
+        let mut resp = self.etcd_rs_client.kv().put(req).await.context(format!(
+            "failed to get PutResponse from etcd for key={:?}, value={:?}",
+            key, value
```
Comment (pwang7) [inline src/etcd_client.rs]: add trailing comma

File: src/etcd_client.rs [hunk 10]
```diff
@@ -83,12 +75,10 @@ impl EtcdClient {
...
         req.set_prev_kv(true);
-        let mut resp = self.etcd_rs_client.kv().delete(req).await.map_err(|e| {
-            anyhow!(
-                "failed to get DeleteResponse from etcd, the error is: {}",
-                e,
-            )
-        })?;
+        let mut resp = self.etcd_rs_client.kv().delete(req).await.context(format!(
+            "failed to get DeleteResponse from etcd for key={:?}",
+            key
```
Comment (pwang7) [inline src/etcd_client.rs]: add trailing comma

**Review by pwang7 — CHANGES_REQUESTED** [review]
please fix test failures

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -38,12 +38,14 @@ jobs:
...
       - uses: actions-rs/toolchain@v1
         with:
-          profile: minimal
           toolchain: ${{ env.CI_RUST_TOOLCHAIN }}
           override: true
       - uses: actions-rs/cargo@v1
         with:
           command: check
+      - name: Setup tmate session
+        if: ${{ always() }}
```
Comment (pwang7) [inline .github/workflows/ci.yml]: no need this here.

File: Cargo.toml [hunk 12]
```diff
@@ -16,7 +16,7 @@ categories = ["filesystem"]
 anyhow = "1.0.31"
 bincode = "1.3.1"
 env_logger = "0.6.0"
-etcd-rs = "0.2.2"
+etcd-rs = { git = "https://github.com/datenlord/etcd-client", rev="3f2feab" }
```
Comment (pwang7) [inline Cargo.toml:19]: etcd-client

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 12 above)
Comment (mrwwheat7) [reply Cargo.toml:19]: We need to change the package name in datenlord/etcd-client, will change it in the next PR

File: src/etcd_client.rs [hunk 13]
```diff
@@ -187,7 +177,7 @@ impl EtcdClient {
                 .kv()
                 .delete(req)
                 .await
-                .map_err(|e| anyhow!("failed to clear all data from etcd, the error is: {}", e,))
+                .context("failed to delete all data from etcd")
```
Comment (pwang7) [inline src/etcd_client.rs]: put `context()` out of `smol::run()`

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -336,7 +334,6 @@ jobs:
       - uses: actions/checkout@v2
       - uses: actions-rs/toolchain@v1
         with:
-          profile: minimal
           toolchain: ${{ env.CI_RUST_TOOLCHAIN }}
           override: true
       - run: rustup component add rustfmt
```
Comment (pwang7) [inline .github/workflows/ci.yml]: change to `profile: default`, then no need to add `rustfmt`

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -353,7 +350,6 @@ jobs:
       - uses: actions/checkout@v2
       - uses: actions-rs/toolchain@v1
         with:
-          profile: minimal
           toolchain: ${{ env.CI_RUST_TOOLCHAIN }}
           override: true
       - run: rustup component add clippy
```
Comment (pwang7) [inline .github/workflows/ci.yml]: change to `profile: default`, then no need to add `clippy`

**Review by pwang7 — CHANGES_REQUESTED** [review]

**Review by pwang7 — DISMISSED** [review]

**Review by pwang7 — APPROVED** [review]

**Pull request merged**